ipa-core/src/protocol/prss/mod.rs

```diff
@@ -60,6 +64,65 @@ impl Debug for UsedSet {
     }
 }
 
+/// Internal PRSS index.
+///
+/// `PrssIndex128` values are directly input to the block cipher used for pseudo-random generation.
+/// Each invocation must use a distinct `PrssIndex128` value. Most code should use the `PrssIndex`
+/// type instead, which often corresponds to record IDs. `PrssIndex128` values are produced by
+/// the `PrssIndex::offset` function and include the primary `PrssIndex` plus a possible offset
+/// when more than 128 bits of randomness are required to generate the requested value.
+///
+/// This is public so that it can be used by the instrumentation wrappers in
+/// `ipa_core::protocol::context`.  It should not generally be used outside the PRSS implementation.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub(crate) struct PrssIndex128(u128);
+
+impl From<u128> for PrssIndex128 {
+    fn from(value: u128) -> Self {
+        Self(value)
+    }
+}
+
+impl From<PrssIndex128> for u128 {
+    fn from(value: PrssIndex128) -> Self {
+        value.0
+    }
+}
+
+/// PRSS index.
+///
+/// PRSS indexes are used to ensure that distinct pseudo-randomness is generated for every value
+/// output by PRSS. It is often sufficient to use record IDs as PRSS indexes, and
+/// `impl From<RecordId> for PrssIndex` is provided for that purpose.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct PrssIndex(u32);
+
+impl From<u32> for PrssIndex {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+
+// It would be nice for this to be TryFrom, but there's a lot of places where we use u128s as PRSS indexes.
+impl From<u128> for PrssIndex {
+    fn from(value: u128) -> Self {
+        Self(value.try_into().unwrap())
+    }
+}
+
+impl From<RecordId> for PrssIndex {
+    fn from(value: RecordId) -> Self {
+        Self(u32::from(value))
+    }
+}
+
+impl PrssIndex {
+    fn offset(self, offset: usize) -> PrssIndex128 {
+        assert!(offset <= u16::MAX.into());
+        PrssIndex128::from(u128::from((u64::from(self.0) << 16) + (offset as u64)))
```

Comment:
Do you really want to only use 16 bits for the offset, given that this is a 32-bit value?

Reply:
Do you think 32 makes more sense? There was also @danielmasny's suggestion of 64 in https://github.com/private-attribution/ipa/issues/884#issuecomment-1854926221.

Implementing `Display` will eliminate one reason for minimizing it, but it also seemed unlikely that any single invocation of PRSS would generate a value with more than 2^23 bits of entropy? I considered as little as 2^15. However, 2^15 bits is a 4,096-byte vector. That is the upper end of the range I foresee using, but I wanted to allow flexibility to go a bit bigger than that.

Reply:
Keeping the values within a 64-bit range is fine, possibly even a good idea.